code/game/machinery/recharger.dm

```diff
@@ -179,6 +146,57 @@
 		return
 	icon_state = icon_state_idle
 
+/obj/machinery/recharger/proc/get_cell_from(obj/item/I)
+	if(istype(I, /obj/item/gun/energy))
+		var/obj/item/gun/energy/E = I
+		return E.cell
+
+	if(istype(I, /obj/item/melee/baton))
+		var/obj/item/melee/baton/B = I
+		return B.cell
+
+	if(istype(I, /obj/item/modular_computer))
+		var/obj/item/modular_computer/C = I
+		var/obj/item/computer_hardware/battery/B = C.all_components[MC_CELL]
+		if(B)
+			return B.battery
+
+	if(istype(I, /obj/item/rcs))
+		var/obj/item/rcs/R = I
+		return R.rcell
+
+	if(istype(I, /obj/item/bodyanalyzer))
+		var/obj/item/bodyanalyzer/B = I
+		return B.cell
+
+	return null
+
+/obj/machinery/recharger/proc/check_cell_needs_recharging(obj/item/stock_parts/cell/C)
+	return C && C.charge < C.maxcharge
+
+/obj/machinery/recharger/proc/recharge_cell(obj/item/stock_parts/cell/C, power_usage)
+	C.give(C.chargerate * recharge_coeff)
+	use_power(power_usage)
+
+/obj/machinery/recharger/proc/try_recharging_if_possible(just_check = FALSE) // just_check means no actual recharging takes place, useful for cosmetic purposes
```

Comment:
I'd split this up in two procs. One that checks if it has to charge and another to actually charge. Which makes use of the prior proc.
This way you ditch the awkward param